### PR TITLE
Refactor article read manager

### PR DESCRIPTION
**Pull request by richtier**
[this task](https://uktrade.atlassian.net/browse/ED-2947)

- "read" can refer to _the action of reading_ or _the state of having been read_ - so it's confusing ot call something a `ReadManager` and methods that use the word `read`. To solve this instead use the words"view" and "viewed"
- the article read manager was too stateful. now it is less stateful.
- the "time to read article" code assumes each row has 5 words. No need to make that assumption.

### CODE REVIEW

**Comment by richtier** [comment]
codecov is failing because an abstractmethod is not called.

File: article/articles.py [hunk 1]
```diff
@@ -21,10 +21,10 @@ def __init__(self, uuid, title, keywords, tasks,
         self.parent = parent
 
     @cached_property
-    def time_to_read(self):
+    def time_to_view(self):
```
Comment (konradko) [inline article/articles.py]: don't love the name, but understand the confusion with 'read' past tense :-)

File: article/helpers.py [hunk 2]
```diff
@@ -7,128 +7,81 @@
...
+    def __init__(self, request):
         self.request = request
-        self.current_article = current_article
-
-    @abc.abstractmethod
-    def retrieve_historic_article_uuids(self):
-        return
 
-    def get_group_read_progress(self):
+    def view_progress_for_groups(self):
```
Comment (konradko) [inline article/helpers.py]: `get_view_progress_for_groups` would be nicer, as this ain't a property

File: article/helpers.py [hunk 3]
```diff
@@ -7,128 +7,81 @@
...
-        # read_articles_in_category is a new set (intersection)
-        # with elements common to read_articles and articles_in_category
-        read_articles_in_group = self.read_article_uuids & article_uuids
-        return read_articles_in_group
-
-    def article_read_count(self, group_name):
-        return len(self.read_articles_keys_in_group(group_name))
+    def articles_viewed_for_group(self, group_name):
+        group = structure.get_article_group(group_name)
+        return self.viewed_article_uuids & group.article_uuids
```
Comment (konradko) [inline article/helpers.py]: please leave the comment about `&` returning intersection

File: article/helpers.py [hunk 4]
```diff
@@ -7,128 +7,81 @@
...
+    def viewed_article_uuids(self):
+        return self.retrieve_viewed_article_uuids()
+
+    @abc.abstractmethod
+    def retrieve_viewed_article_uuids(self):
+        return
 
 
-class ArticleReadManager:
+class ArticlesViewedManager:
```
Comment (konradko) [inline article/helpers.py]: let's rename the model in directory-api to `ArticleViewed` as well

Comment (konradko) [reply article/helpers.py]: in terms of patterns, this is rather `ArticlesViewedManagerFactory`

File: article/helpers.py [hunk 5]
```diff
@@ -7,128 +7,81 @@
...
     def __new__(cls, request, current_article=None):
-        session_manager = SessionArticlesReadManager(request, current_article)
-        database_manager = DatabaseArticlesReadManager(
-            request, current_article
-        )
+        session_manager = SessionArticlesViewedManager(request)
```
Comment (konradko) [inline article/helpers.py]: it is a bit of unexpected that instantiating a class can cause DB reads and writes. But I understand this is to handle the complexity of syncing session and db progress.

To keep it explicit let's split the code below into 2 paths - user logged in and not (e.g.

```
if request.sso_user is None:
    articles_viewed_manager = self.get_session_articles_viewed_manager(request, current_article)
else:
    articles_viewed_manager = self.get_database_articles_viewed_manager(request, current_article)

return articles_viewed_manager
```

File: article/helpers.py [hunk 6]
```diff
@@ -7,128 +7,81 @@
...
+
         if request.sso_user is None:
             if current_article:
                 session_manager.persist_article(current_article.uuid)
             return session_manager
-        response = database_manager.bulk_persist_article(
-            article_uuids=session_manager.read_article_uuids
-        )
+
+        viewed_article_uuids = set()
```
Comment (konradko) [inline article/helpers.py]: would be good to comment on what this achieves or move it to a function with an explicit name (e.g. `sync_session_and_database_viewed_articles`)

File: article/articles.py (hunk 1 above)
Comment (konradko) [reply article/articles.py]: as discussed, let's keep it as `time_to_read` while changing ArticleRead to ArticleViewed

File: article/structure.py [hunk 7]
```diff
@@ -12,16 +12,22 @@ def __init__(self, articles, name, title, url, next_guidance_group=None):
...
         self.url = url
         self.next_guidance_group = next_guidance_group
-        self.articles_set = frozenset(
+        self.article_uuids = frozenset(
             [article.uuid for article in self.articles]
         )
 
     @cached_property
-    def total_reading_time(self):
+    def view_time(self):
```
Comment (konradko) [inline article/structure.py]: let's use consistent `read_time` and `view_progress`

File: article/tests/test_helpers.py [hunk 8]
```diff
@@ -40,8 +40,8 @@ def test_database_create_article_read_calls_api(
 ):
     mock_bulk_create_article_read.return_value = create_response(200)
 
-    manager = helpers.DatabaseArticlesReadManager(sso_request)
-    manager.bulk_persist_article(article_uuids=['123'])
+    manager = helpers.DatabaseArticlesViewedManager(sso_request)
+    manager.persist_articles(article_uuids=['123'])
```
Comment (konradko) [inline article/tests/test_helpers.py]: `save_article_views` or `persist_article_views` would be a better name, as it's the views, not the articles being persisted here.

File: article/tests/test_helpers.py [hunk 9]
```diff
@@ -197,36 +196,36 @@ def test_article_read_manager_handles_no_read_articles_on_synchronisation(
...
-        helpers.ArticleReadManager(sso_request)
+        helpers.ArticlesViewedManagerFactory(sso_request)
 
     assert sso_request.session[session_key] == [1, 2, 3]
     assert str(execinfo.value) == 'validation error'
 
 
 @patch('api_client.api_client.exportreadiness.bulk_create_article_read')
-def test_database_remaining_reading_time_in_group(
+def test_database_remaining_view_time_for_group(
```
Comment (konradko) [inline article/tests/test_helpers.py]: `view_time` > `read_time`

**Pull request merged**